### PR TITLE
Invalidate all feature, refs #690

### PR DESCRIPTION
**Pull request by batpad**
This adds a button in Project Edit / Misc for admins to mark all "done" tasks on the project as invalid.

See #690 for discussions on the issue.

Please let me know if something can be improved. Thank you.

### CODE REVIEW

**Comment by batpad** [comment]
@pgiraud my apologies - had some pep8 errors in there. Should be all fixed now. Let me know if anything else is needed from my end here. Take your time on merging - it is possible that this feature requires further discussion, etc.

File: osmtm/static/js/project.edit.js [hunk 1]
```diff
@@ -191,6 +191,41 @@ osmtm.project.edit.priority_areas = (function() {
...
+      setTimeout(function() {
+        $success.fadeOut();
+      }, 5000);
+    } else {
+      var error = data.error_msg;
+      $errors.text(error);
+      $errors.prepend('<span class="glyphicon glyphicon-exclamation-sign"></span> ');
+      // $('#task_error_msg').html(data.error_msg).show()
+      //   .delay(3000)
+      //   .fadeOut();
```
Comment (pgiraud) [inline osmtm/static/js/project.edit.js]: If not useful, could you please remove those comment lines?

**Comment by batpad** [comment]
@pgiraud Apologies - thanks for the catch. Removed.

File: osmtm/templates/project.edit.mako [hunk 2]
```diff
@@ -36,6 +36,43 @@
...
+        <h4 class="modal-title" id="myModalLabel">${_('Invalidate All')}</h4>
+      </div>
+      <div class="modal-body">
+        <p>
+          ${_("This will mark all tasks currently marked as 'done' as invalid. Please use this only if you are sure of what you are doing.")}
+        </p>
+        <div class="text-center">
+          <div class="form-group">
+            <textarea id="project_invalidate_comment" name="invalidate_all_comment" class="form-control" placeholder="${_('Leave a comment')}" rows="2"></textarea>
+
```
Comment (pgiraud) [inline osmtm/templates/project.edit.mako]: Is there any reason why you leave a blank line here?

File: osmtm/templates/task.empty.mako [hunk 3]
```diff
@@ -29,3 +29,17 @@
...
+  <div class="form-group">
+    <textarea id="project_invalidate_comment" name="invalidate_all_comment" class="form-control" placeholder="${_('Leave a comment')}" rows="2"></textarea>
+
+  </div>
+  <a class="btn btn-danger btn-lg btn-invalidate-all">
+     <span class="glyphicon glyphicon-share-alt"></span>&nbsp;
+     ${_('Invalidate all done tasks')}
+  </a>
+</div>
+% endif
```
Comment (pgiraud) [inline osmtm/templates/task.empty.mako]: I'm not in favor of adding the invalidate button elsewhere than in the project edit page.
Putting it in the project main page would tempt the project managers to use it even if they don't know that it's an advanced feature.
Furthermore it didn't do anything when I tried to use it.

File: osmtm/templates/project.edit.mako [hunk 4]
```diff
@@ -36,6 +36,43 @@
...
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        <h4 class="modal-title" id="myModalLabel">${_('Invalidate All')}</h4>
+      </div>
+      <div class="modal-body">
+        <p>
+          ${_("This will mark all tasks currently marked as 'done' as invalid. Please use this only if you are sure of what you are doing.")}
+        </p>
+        <div class="text-center">
+          <div class="form-group">
+            <textarea id="project_invalidate_comment" name="invalidate_all_comment" class="form-control" placeholder="${_('Leave a comment')}" rows="2"></textarea>
```
Comment (pgiraud) [inline osmtm/templates/project.edit.mako]: I had to read the code to understand what the comment field was used for. It wasn't obvious to me. I thought that it was a way to ensure that the project manager did really want to do the invalidation. I was tempted to fill it with random text.
I think that this is important for the user to understand that this comment will appear in the history of each task concerned by the invalidation. We can change the placeholder, add a help text below the comment textarea or even write a default comment.

File: osmtm/templates/project.edit.mako [hunk 5]
```diff
@@ -36,6 +36,43 @@
...
+
+          </p>
+          <a class="btn btn-danger btn-lg btn-invalidate-all">
+             <span class="glyphicon glyphicon-share-alt"></span>&nbsp;
+             ${_('Invalidate all done tasks')}
+          </a>
+        </div>
+      </div>
+      <div class="modal-footer">
+        <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
```
Comment (pgiraud) [inline osmtm/templates/project.edit.mako]: I think that "Cancel" is better than "Close" here.

File: osmtm/views/project.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
     TaskState,
     TaskLock,
     License,
+    INVALIDATED,
```
Comment (pgiraud) [inline osmtm/views/project.py]: I generally prefer not to use the constant directly. I use the class member instead. ie. `Task.state_invalidated`.

File: osmtm/views/project.py [hunk 7]
```diff
@@ -493,6 +494,65 @@ def project_preset(request):
...
+    project = DBSession.query(Project).get(id)
+    user_id = authenticated_userid(request)
+    user = DBSession.query(User).get(user_id)
+    tasks = project.tasks
+    tasks_affected = 0
+    for task in tasks:
+        if task.cur_state.state == TaskState.state_done:
+            tasks_affected += 1
+            task.user = None
+            task.states.append(TaskState(user=user, state=INVALIDATED))
```
Comment (pgiraud) [inline osmtm/views/project.py]: `Task.state_invalidated` preferred.

File: osmtm/tests/test_project.py [hunk 8]
```diff
@@ -727,3 +727,86 @@ def test_project_check_expiration(self):
...
+        # set correct challenge id
+        params['challenge_id'] = project_id
+
+        self.testapp.post('/project/%d/invalidate_all' % project_id,
+                          params=params, headers=headers,
+                          xhr=True, status=200)
+
+        # check that our done tasks are invalidated
+        task0 = DBSession.query(Task).get((project_id, tasks[0].id))
+        self.assertEqual(task0.cur_state.state, INVALIDATED)
```
Comment (pgiraud) [inline osmtm/tests/test_project.py]: `Task.state_invalidated` preferred here.

File: osmtm/tests/test_project.py [hunk 9]
```diff
@@ -727,3 +727,86 @@ def test_project_check_expiration(self):
...
+        task3 = DBSession.query(Task).get((project_id, tasks[3].id))
+        self.assertEqual(task3.cur_state.state, INVALIDATED)
```
Comment (pgiraud) [inline osmtm/tests/test_project.py]: `Task.state_invalidated` preferred here.

File: osmtm/tests/test_project.py [hunk 10]
```diff
@@ -727,3 +727,86 @@ def test_project_check_expiration(self):
...
+
+        # check that other tasks are not touched
+        task1 = DBSession.query(Task).get((project_id, tasks[1].id))
+        self.assertEqual(task1.cur_state.state, READY)
```
Comment (pgiraud) [inline osmtm/tests/test_project.py]: `Task.state_ready` preferred here.

**Comment by pgiraud** [comment]
I added some minor comments.
Otherwise this looks good to me.
Please address the comments if you agree.

**Comment by batpad** [comment]
wow, thank you so much for the detailed review here, @pgiraud !

I agree with everything here. Some of this was my carelessness (apologies), and some not knowing project conventions. I will fix these up as soon as possible and update the PR, and make sure to be more diligent in the future. Thanks again for the detailed review.

File: osmtm/templates/project.edit.mako (hunk 2 above)
Comment (batpad) [inline osmtm/templates/project.edit.mako]: Fixed as part of 4720f71

File: osmtm/templates/task.empty.mako (hunk 3 above)
Comment (batpad) [inline osmtm/templates/task.empty.mako]: This was an over-sight on my part. Fixed in https://github.com/osm-in/osm-tasking-manager2/commit/be443d2020f3df6b15e15131ce3fcb3b91a5a0ea

File: osmtm/templates/project.edit.mako (hunk 4 above)
Comment (batpad) [inline osmtm/templates/project.edit.mako]: I made some changes here: https://github.com/osm-in/osm-tasking-manager2/commit/4720f71438acc4bcaaa55223f7caf15c3272c198

I am not convinced that this is still ideal. Would appreciate more specific ideas to how to make this look nicer and be clearer in what it does.

File: osmtm/templates/project.edit.mako (hunk 5 above)
Comment (batpad) [inline osmtm/templates/project.edit.mako]: Makes sense. Done here: https://github.com/osm-in/osm-tasking-manager2/commit/323396c467d3603377acc12354f8a72a29439aa4

File: osmtm/views/project.py (hunk 6 above)
Comment (batpad) [inline osmtm/views/project.py]: Not importing constants directly and using properties of TaskState instead: https://github.com/hotosm/osm-tasking-manager2/commit/d8c8ce608719206a39cc8c6db3246322ed7454e6

**Comment by pgiraud** [comment]
Hi @batpad, you did a really great work.
I worked a little bit on the layout of the modal window trying to make it look more clear.

You can test the result by checkouting the branch I worked in
https://github.com/hotosm/osm-tasking-manager2/compare/master...pgiraud:invalidate-all
Here's the commit to cherry-pick if you agree with my modifications:
https://github.com/pgiraud/osm-tasking-manager2/commit/887d2b73ba7f7dda281387dd2aed59d776c5cfed

**Comment by batpad** [comment]
@pgiraud the smaller modal with much clearer text looks great! Have cherry-picked and pushed.

It was a really good experience for me working with you on this feature. I hope to be able to continue to contribute to this wonderful project.

**Pull request merged**